src/frontend/src/optimizer/rule/pull_up_correlated_predicate_agg_rule.rs

```diff
@@ -0,0 +1,245 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use itertools::{Either, Itertools};
+use risingwave_common::types::DataType;
+use risingwave_common::util::column_index_mapping::ColIndexMapping;
+use risingwave_expr::aggregate::AggKind;
+
+use super::super::plan_node::*;
+use super::{BoxedRule, Rule};
+use crate::expr::{Expr, ExprImpl, ExprRewriter, ExprType, FunctionCall, InputRef};
+use crate::optimizer::plan_node::generic::{Agg, GenericPlanNode, GenericPlanRef};
+use crate::optimizer::plan_visitor::{PlanCorrelatedIdFinder, PlanVisitor};
+use crate::optimizer::PlanRef;
+use crate::utils::{Condition, IndexSet};
+
+/// Pull up correlated predicates from the right agg side of Apply to the `on` clause of Join.
+///
+/// Before:
+///
+/// ```text
+///     LogicalApply
+///    /            \
+///  LHS          Project
+///                 |
+///                Agg [group by nothing]
+///                 |
+///               Project
+///                 |
+///               Filter [correlated_input_ref(yyy) = xxx]
+/// ```
+///
+/// After:
+///
+/// ```text
+///     LogicalApply [yyy = xxx]
+///    /            \
+///  LHS          Project
+///                 |
+///                Agg [group by xxx]
+///                 |
+///               Project
+///                 |
+///               Filter
+/// ```
+
+pub struct PullUpCorrelatedPredicateAggRule {}
+impl Rule for PullUpCorrelatedPredicateAggRule {
+    fn apply(&self, plan: PlanRef) -> Option<PlanRef> {
+        let apply = plan.as_logical_apply()?;
+        let (apply_left, apply_right, apply_on, join_type, correlated_id, _, max_one_row) =
+            apply.clone().decompose();
+
+        if max_one_row {
+            return None;
+        }
+
+        let top_project = if let Some(project) = apply_right.as_logical_project() {
+            project.clone()
+        } else {
+            LogicalProject::with_mapping(
+                apply_right.clone(),
+                ColIndexMapping::identity(apply_right.schema().len()),
+            )
+        };
+        let (top_proj_exprs, _) = top_project.clone().decompose();
+
+        let input = top_project.input();
+        let agg: &LogicalAgg = input.as_logical_agg()?;
+        let (agg_calls, group_key, grouping_sets, input, _enable_two_phase) =
+            agg.clone().decompose();
+        // It could be too restrictive to require the group key to be empty. We can relax this in the future if necessary.
+        if !group_key.is_empty() {
+            return None;
+        }
```

Comment:
is simply adding tge correlated key(`xxx`) into the group key correct?

Reply:
No, simply adding those group keys is not correct. We need to handle the `new_agg` parent input reference in a more sophisticated way instead of the current simple shifting.